magic_items/routes.py

```diff
@@ -0,0 +1,304 @@
+from magic_items.services import (
+    MagicItemService,
+    ItemForm,
+    ItemFilterForm,
+    CollectionAddForm,
+    CollectionService,
+    ItemCollectionService,
+)
+from users.services import UserService
+
+magic_routes = Blueprint("magic_routes", __name__)
+
+
+def before_manage_magic_items(func):
+    def wrapper(*args, **kwargs):
+        if not g.user:
+            flash("Please login or signup to manage magic items", "warning")
+            return redirect("/login")
+        return func(*args, **kwargs)
+
+    wrapper.__name__ = func.__name__ + wrapper.__name__
+    return wrapper
+
+
+def before_manage_collections(func):
+    def wrapper(*args, **kwargs):
+        if not g.user:
+            flash("Please login or signup to manage collections", "warning")
+            return redirect("/login")
+        return func(*args, **kwargs)
+
+    wrapper.__name__ = func.__name__ + wrapper.__name__
+    return wrapper
+
+
+#########################################################################################################
+#########################################################################################################
+##
+## Magic Item Routes
+
+
+@magic_routes.route("/magic-items/new", methods=["GET", "POST"])
+@before_manage_magic_items
+def add_item():
+
+    form = ItemForm()
+
+    if form.validate_on_submit():
+        try:
+            item = MagicItemService.create(g.user.id, form)
+            flash("new item " + item.name + " added", "success")
+            return redirect(f"/magic-items/{item.id}")
+
+        except:
+            flash("error", "danger")
+            return render_template("magic_items/new_magic_item.html", form=form)
+
+    return render_template("magic_items/new_magic_item.html", form=form)
+
+
+@magic_routes.route("/magic-items/<int:item_id>", methods=["GET"])
+def show_item(item_id):
+    item = MagicItemService.get(item_id)
+    if item.created_by:
+        user = UserService.get_by_id(item.created_by)
+        return render_template(
+            "magic_items/magic_item.html",
+            item=item,
+            user=user,
+        )
+    return render_template(
+        "magic_items/magic_item.html",
+        item=item,
+    )
+
+
+@magic_routes.route("/magic-items", methods=["GET"])
+def show_items():
+
+    form = ItemFilterForm(object=request.args)
+
+    if "created_by" in request.args and request.args["created_by"] == "":
+        flash("Please login to manage magic items", "warning")
+        return redirect("/login")
+
+    items = MagicItemService.get_filtered(**request.args)
+
+    return render_template(
+        "magic_items/magic_items.html",
+        items=items,
+        form=form,
+    )
+
+
+@magic_routes.route("/magic-items/edit", methods=["GET", "POST"])
+@before_manage_magic_items
+def edit_item():
+
+    magic_item_id = int(request.args["magic_item_id"])
+    item = MagicItemService.get(magic_item_id)
```

Comment:
Not sure why we need to fetch the item first before editing it. The usual pattern is to pass in the id of the resource being updated and the changes.

Reply:
Currently the magic is fetched here for 3 uses:

- feed it's data into the form so the user does not need to start from scratch
- send the object to the edit_magic_item.html template
- grab the second description item in the item list

Is there a better way to solve the above problems, or should I adjust/document the code to make it clearer?